src/ThreadPool.cc

```diff
@@ -0,0 +1,327 @@
+    std::function<void(ThreadPool&)> fct, TimePoint start_time, std::string name)
+{
+    if (!fct)
+        throw std::invalid_argument("ThreadPool does not accept null functions");
+
+    auto task_id = [this, &fct, start_time, &name]()
+        {
+            std::lock_guard<std::mutex> lock(mutex_);
+
+            if (is_full_i())
+            {
+                throw std::runtime_error(
+                    cat("Cannot add task: Pending queue has reached capacity (",
+                        pending_tasks_.size(), ")"));
+            }
+
+            const auto id = next_task_id_;
+            pending_tasks_.emplace_back(id, std::move(fct), std::move(name), start_time);
+
+            ++next_task_id_;
+
+            return id;
+        }();
+
+    cv_.notify_one();
+
+    return task_id;
+}
+
+ThreadPool::TaskId ThreadPool::add_task(
+    std::function<void()> fct, TimePoint start_time, std::string name)
+{
+    if (!fct)
+        throw std::invalid_argument("ThreadPool does not accept null functions");
+
+    return add_task(
+        [f = std::move(fct)](ThreadPool&) { f(); }, start_time, std::move(name));
+}
+
+ThreadPool::TaskId ThreadPool::add_task(std::function<void(ThreadPool&)> fct,
+    ThreadPool::Duration delay_before_start, std::string name)
+{
+    return add_task(std::move(fct),
+        std::chrono::system_clock::now() + delay_before_start, std::move(name));
+}
+
+ThreadPool::TaskId ThreadPool::add_task(std::function<void()> fct,
+    ThreadPool::Duration delay_before_start, std::string name)
+{
+    return add_task(std::move(fct),
+        std::chrono::system_clock::now() + delay_before_start, std::move(name));
+}
+
+ThreadPool::TaskId ThreadPool::add_task(
+    std::function<void(ThreadPool&)> fct, std::string name)
+{
+    return add_task(std::move(fct), TimePoint{}, std::move(name));
+}
+
+ThreadPool::TaskId ThreadPool::add_task(std::function<void()> fct, std::string name)
+{
+    return add_task(std::move(fct), TimePoint{}, std::move(name));
+}
+
+std::size_t ThreadPool::count_pending() const
+{
+    std::lock_guard<std::mutex> lock(mutex_);
+    return count_pending_i();
+}
+
+std::size_t ThreadPool::count_pending_i() const
+{
+    return pending_tasks_.size();
+}
+
+std::size_t ThreadPool::count_threads() const noexcept
+{
+    return threads_.size();
+}
+
+/// Return a vector with the IDs of the jobs that are currently running.
+std::vector<ThreadPool::TaskId> ThreadPool::get_running_task_ids() const
+{
+    std::lock_guard<std::mutex> lock(mutex_);
+    return running_task_ids_;
+}
+
+/// Return a vector with the names of the jobs that are currently running.
+std::vector<std::string> ThreadPool::get_running_task_names() const
+{
+    std::lock_guard<std::mutex> lock(mutex_);
+    return running_task_names_;
+}
+
+bool ThreadPool::is_empty() const noexcept
+{
+    std::lock_guard<std::mutex> lock(mutex_);
+    return is_empty_i();
+}
+
+bool ThreadPool::is_empty_i() const noexcept
```

Comment:
private `is_empty_i()` seems to be not used anywhere, why do we have it this way?

`is_full_i()` otoh is used in internal context where we want to avoid locking.

Reply:
Good catch! I guess they were used in a previous implementation. I have thrown them out.